File: Base64Encoder.cs
// Base64Encoder.cs
using System;
using System.IO;
using System.Text;
using System.Threading;

class Base64Encoder
{
    private const int ChunkSize = 1024 * 1024;

    static void ProgressBar(long current, long total, int width = 40)
    {
        if (total == 0) return;
        double percent = (double)current / total;
        int filled = (int)(width * percent);
        string bar = new string('█', filled) + new string('░', width - filled);
        Console.Error.Write($"\rProgress: [{bar}] {percent * 100:F1}%");
        if (current >= total) Console.Error.WriteLine();
    }

    static void EncodeFile(string inputPath, string outputPath, bool urlSafe)
    {
        if (!File.Exists(inputPath))
        {
            Console.Error.WriteLine($"Error: input file '{inputPath}' not found.");
            Environment.Exit(1);
        }
        var fileInfo = new FileInfo(inputPath);
        long fileSize = fileInfo.Length;
        var outStream = outputPath != null ? File.Create(outputPath) : Console.OpenStandardOutput();
        using var inStream = File.OpenRead(inputPath);
        byte[] buffer = new byte[ChunkSize];
        int read;
        long processed = 0;
        while ((read = inStream.Read(buffer, 0, buffer.Length)) > 0)
        {
            string encoded = urlSafe ?
                Convert.ToBase64String(buffer, 0, read).Replace('+', '-').Replace('/', '_') :
                Convert.ToBase64String(buffer, 0, read);
            byte[] bytes = Encoding.ASCII.GetBytes(encoded);
            outStream.Write(bytes, 0, bytes.Length);
            processed += read;
            ProgressBar(processed, fileSize);
        }
        if (outputPath != null) outStream.Close();
        Console.WriteLine($"\n✅ Encoded '{inputPath}' to '{(outputPath ?? "stdout")}'");
    }

    static void DecodeFile(string inputPath, string outputPath, bool urlSafe)
    {
        if (!File.Exists(inputPath))
        {
            Console.Error.WriteLine($"Error: input file '{inputPath}' not found.");
            Environment.Exit(1);
        }
        string data = File.ReadAllText(inputPath).Replace(" ", "").Replace("\n", "").Replace("\r", "").Replace("\t", "");
        // Add padding
        int missing = data.Length % 4;
        if (missing > 0) data += new string('=', 4 - missing);
        byte[] decoded;
        try
        {
            if (urlSafe)
            {
                string standard = data.Replace('-', '+').Replace('_', '/');
                decoded = Convert.FromBase64String(standard);
            }
            else
            {
                decoded = Convert.FromBase64String(data);
            }
        }
        catch (Exception e)
        {
            Console.Error.WriteLine($"Error: invalid Base64 input - {e.Message}");
            Environment.Exit(1);
            return;
        }
        var outStream = outputPath != null ? File.Create(outputPath) : Console.OpenStandardOutput();
        outStream.Write(decoded, 0, decoded.Length);
        if (outputPath != null) outStream.Close();
        Console.WriteLine($"\n✅ Decoded '{inputPath}' to '{(outputPath ?? "stdout")}'");
    }

    static void Main(string[] args)
    {
        string encode = null, decode = null, output = null;
        bool urlSafe = false;
        for (int i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "-e": encode = args[++i]; break;
                case "-d": decode = args[++i]; break;
                case "-o": output = args[++i]; break;
                case "-u": urlSafe = true; break;
                case "-h": case "--help":
                    Console.WriteLine("Usage: dotnet run -- -e <file> [-o <out>] [-u]\n       dotnet run -- -d <file> [-o <out>] [-u]");
                    return;
            }
        }
        if (encode == null && decode == null)
        {
            Console.Error.WriteLine("Error: either -e or -d required");
            Environment.Exit(1);
        }
        if (encode != null) EncodeFile(encode, output, urlSafe);
        else DecodeFile(decode, output, urlSafe);
    }
}

File: Base64Encoder.java
// Base64Encoder.java
import java.io.*;
import java.nio.file.*;
import java.util.Base64;

public class Base64Encoder {
    private static final int CHUNK_SIZE = 1024 * 1024;

    private static void progressBar(long current, long total, int width) {
        if (total == 0) return;
        double percent = (double) current / total;
        int filled = (int) (width * percent);
        StringBuilder bar = new StringBuilder();
        for (int i = 0; i < filled; i++) bar.append('█');
        for (int i = filled; i < width; i++) bar.append('░');
        System.err.printf("\rProgress: [%s] %.1f%%", bar, percent * 100);
        if (current >= total) System.err.println();
    }

    private static void encodeFile(String inputPath, String outputPath, boolean urlSafe) throws IOException {
        Path input = Paths.get(inputPath);
        if (!Files.exists(input)) {
            System.err.println("Error: input file '" + inputPath + "' not found.");
            System.exit(1);
        }
        long fileSize = Files.size(input);
        OutputStream out = outputPath != null ? Files.newOutputStream(Paths.get(outputPath)) : System.out;
        Base64.Encoder encoder = urlSafe ? Base64.getUrlEncoder() : Base64.getEncoder();
        try (InputStream in = Files.newInputStream(input)) {
            byte[] buffer = new byte[CHUNK_SIZE];
            int read;
            long processed = 0;
            while ((read = in.read(buffer)) != -1) {
                byte[] encoded = encoder.encode(buffer, 0, read);
                out.write(encoded);
                processed += read;
                progressBar(processed, fileSize, 40);
            }
        }
        if (outputPath != null) out.close();
        System.out.printf("\n✅ Encoded '%s' to '%s'\n", inputPath, outputPath != null ? outputPath : "stdout");
    }

    private static void decodeFile(String inputPath, String outputPath, boolean urlSafe) throws IOException {
        Path input = Paths.get(inputPath);
        if (!Files.exists(input)) {
            System.err.println("Error: input file '" + inputPath + "' not found.");
            System.exit(1);
        }
        String data = new String(Files.readAllBytes(input)).replaceAll("\\s", "");
        // Add padding
        int missing = data.length() % 4;
        if (missing > 0) data += "=".repeat(4 - missing);
        byte[] decoded;
        try {
            Base64.Decoder decoder = urlSafe ? Base64.getUrlDecoder() : Base64.getDecoder();
            decoded = decoder.decode(data);
        } catch (IllegalArgumentException e) {
            System.err.println("Error: invalid Base64 input - " + e.getMessage());
            System.exit(1);
            return;
        }
        OutputStream out = outputPath != null ? Files.newOutputStream(Paths.get(outputPath)) : System.out;
        out.write(decoded);
        if (outputPath != null) out.close();
        System.out.printf("\n✅ Decoded '%s' to '%s'\n", inputPath, outputPath != null ? outputPath : "stdout");
    }

    public static void main(String[] args) throws IOException {
        String encode = null, decode = null, output = null;
        boolean urlSafe = false;
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-e": encode = args[++i]; break;
                case "-d": decode = args[++i]; break;
                case "-o": output = args[++i]; break;
                case "-u": urlSafe = true; break;
                case "-h":
                case "--help":
                    System.out.println("Usage: java Base64Encoder -e <file> [-o <out>] [-u]\n       java Base64Encoder -d <file> [-o <out>] [-u]");
                    System.exit(0);
            }
        }
        if (encode == null && decode == null) {
            System.err.println("Error: either -e or -d required");
            System.exit(1);
        }
        if (encode != null) encodeFile(encode, output, urlSafe);
        else decodeFile(decode, output, urlSafe);
    }
}
